File: c++/0896.cpp
// Link: https://leetcode.com/problems/monotonic-array/description/
// Runtime: 123 ms
// Memory: 96.8 MB

class Solution {
public:
  bool isMonotonic(const std::vector<int> &nums) {
    if (nums.size() < 3)
      return true;

    short direction = 0;
    for (int i = 1; i < static_cast<int>(nums.size()); i++) {
      if ((nums[i - 1] < nums[i]) && (direction == 0))
        direction = 1;
      else if ((nums[i - 1] < nums[i]) && (direction == -1))
        return false;
      else if ((nums[i - 1] > nums[i]) && (direction == 0))
        direction = -1;
      else if ((nums[i - 1] > nums[i]) && (direction == 1))
        return false;
    }

    return true;
  }
};

File: python/0896.py
# Link: https://leetcode.com/problems/monotonic-array/description/
# Runtime: 868 ms
# Memory: 30.4 MB


class Solution:
    def isMonotonic(self, nums: List[int]) -> bool:
        if len(nums) < 3:
            return True

        direction = 0
        for i in range(1, len(nums)):
            if (nums[i - 1] < nums[i]) and (direction == 0):
                direction = 1
            elif (nums[i - 1] < nums[i]) and (direction == -1):
                return False
            elif (nums[i - 1] > nums[i]) and (direction == 0):
                direction = -1
            elif (nums[i - 1] > nums[i]) and (direction == 1):
                return False

        return True
